velox/serializers/UnsafeRowSerializer.cpp

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/serializers/UnsafeRowSerializer.h"
+#include "velox/row/UnsafeRowDeserializer.h"
+#include "velox/row/UnsafeRowDynamicSerializer.h"
+
+namespace facebook::velox::serializer::spark {
+
+void UnsafeRowVectorSerde::estimateSerializedSize(
+    VectorPtr vector,
+    const folly::Range<const IndexRange*>& ranges,
+    vector_size_t** sizes) {
+  for (auto& range : ranges) {
+    for (auto i = range.begin; i < range.begin + range.size; ++i) {
+      size_t rowSize = velox::row::UnsafeRowDynamicSerializer::getSize(
+          vector->type(), vector, i);
+      (*sizes)[i] += rowSize;
+    }
+  }
+}
+
+namespace {
+class UnsafeRowVectorSerializer : public VectorSerializer {
+ public:
+  UnsafeRowVectorSerializer(StreamArena* streamArena)
+      : streamArena_{streamArena} {}
+
+  void append(
+      RowVectorPtr vector,
+      const folly::Range<const IndexRange*>& ranges) override {
+    size_t totalSize = 0;
+    for (auto& range : ranges) {
+      for (auto i = range.begin; i < range.begin + range.size; ++i) {
+        size_t rowSize = velox::row::UnsafeRowDynamicSerializer::getSize(
+            vector->type(), vector, i);
+        totalSize += rowSize + sizeof(size_t);
+      }
+    }
+
+    if (totalSize == 0) {
+      return;
+    }
+
+    // Try to allocate 'totalSize' bytes, but handle cases where allocated size
+    // is less.
+    ByteRange byteRange;
+    streamArena_->newRange(totalSize, &byteRange);
+    VELOX_CHECK_EQ(byteRange.position, 0);
+
+    size_t offset = 0;
+    for (auto& range : ranges) {
+      for (auto i = range.begin; i < range.begin + range.size; ++i) {
+        // Write row data.
+        size_t rowSize = velox::row::UnsafeRowDynamicSerializer::getSize(
+            vector->type(), vector, i);
+
+        // Check if there is enough space in the buffer. If not, allocate a new
+        // buffer.
+        if (byteRange.size < offset + rowSize + sizeof(size_t)) {
```

Comment:
For the first ```byteRange``` the ```size``` is gonna be set to ```totalSize```. Given the deterministic nature of the size function this situation is unlikely to happen I guess.

Reply:
> streamArena_->newRange(totalSize, &byteRange);

This call is not guaranteed to allocate totalSize bytes. It may allocate less. Hence, we need to handle the case when first allocation didn't allocate enough data. Without the extra logic the test fails intermittently.